docs/hub/how-to-upstream.md

```diff
@@ -0,0 +1,259 @@
+>>> folder = HfFolder()
+>>> token = folder.get_token()
+```
+
+2. Give `delete_repo` your token and the name of the repository to delete:
+
+```python
+>>> api.delete_repo(token=YOUR_HF_API_TOKEN name=REPO_NAME)
+```
+
+Delete a dataset repository by adding the `type` parameter:
+
+```python
+>>> api.create_repo(token=YOUR_HF_API_TOKEN, name=REPO_NAME, type="dataset")
+```
+
+### Change repository visibility
+
+A repository can be public or private. A private repository is only visible to you or members of your organization. Change a repository to private as shown in the following:
+
+```python
+>>> api.update_repo_visibility(token=YOUR_HF_API_TOKEN, name=REPO_NAME, private=True)
+```
+
+## `Repository` 
+
+The `Repository` class allows you to push models or other repositories to the Hub. `Repository` is a wrapper over Git and Git-LFS methods, so make sure you have Git-LFS installed (see [here](https://git-lfs.github.com/) for more instructions) and set up before you begin. If you are already familiar with common Git commands, then the `Repository` class should feel familiar. 
+
+Start by instantiating a `Repository` object with a path to a local Git clone or repository:
+
+```python
+>>> from huggingface_hub import Repository
+>>> repo = Repository(local_dir="<path>/<to>/<folder>")
+```
+
+### Clone a repository
+
+The `clone_from` parameter clones a repository from a specified directory using a URL (if you are working offline, this parameter should be `None`):
+
+```python
+>>> repo = Repository(local_dir="huggingface-hub", clone_from="https://github.com/huggingface/huggingface_hub")
+```
+
+This parameter can also clone a repository from a Hugging Face model ID:
+
+```python
+>>> repo = Repository(local_dir="w2v2", clone_from="facebook/wav2vec2-large-960h-lv60")
+```
+
+If you want to commit or push to a cloned repository that belongs to you or your organizations:
+
+1. Log in to your Hugging Face account with the following command:
+
+   ```bash
+   huggingface-cli login
+   ```
+
+2. Instantiate your `Repository` class with the parameter `use_auth_token=True`:
+   
+   ```python
+   >>> repo = Repository(local_dir="my-model", clone_from="<user>/<model_id>", use_auth_token=True)
+   ```
+
+You can also attribute a Git username and email to a cloned repository by specifiying the `git_user` and `git_email` parameters. When users commit to that repository, Git will be aware of the commit author.
+
+```python
+>>> repo = Repository(
+...   "my-dataset", 
+...   clone_from="<user>/<dataset_id>", 
+...   use_auth_token=True, 
+...   repo_type="dataset",
+...   git_user="MyName",
+...   git_email="me@cool.mail"
+... )
+```
+
+### Branch
+
+Switch between branches with `git_checkout`. For example, if you want to switch from `branch1` to `branch2`:
+
+```python
+>>> repo = Repository(local_dir="huggingface-hub", clone_from="<user>/<dataset_id>", revision='branch1')
+>>> repo.git_checkout("branch2")
+```
+
+### Add
```

Comment:
To be honest I've never seen people using the add, push and commit methods. People usually use the `push_to_hub` wrapper method. I would probably bias towards simplifying our documentation, but maybe there are use cases I'm missing. WDYT @LysandreJik ?

Reply:
I agree with you, and highlighted something along those lines in my review as well: https://github.com/huggingface/huggingface_hub/pull/358#discussion_r713492640

Reply:
I think if no one is really using the individual `git_add`, `git_push`, and `git_commit` functions then I'll take them out, especially if the `push_to_hub` and `commit` context manager handles these. I will highlight the `commit` context manager first since it is more intuitive, and clarify that `push_to_hub` requires two extra steps first (pull, and save).